devices/r4isdhc.cpp

```diff
@@ -217,29 +279,28 @@ class R4iSDHC : Flashcart {
     }
 
     bool initialize() {
-        CmdBuf4 buf;
-        // this is actually the NOR write disable command
-        // the r4isdhc will respond to cart commands with 0xFFFFFFFF if
-        // the "magic" command hasn't been sent, so we check for that
-        sendCommand(0x40199, 4, buf.u8, 0x180000);
-        if (buf.u32 != 0xFFFFFFFF) {
-            return false;
-        }
-
-        ntrcard::init();
-        sendCommand(0x68, 4, nullptr, 0x180000);
-        // now it will return zeroes
-        sendCommand(0x40199, 4, buf.u8, 0x180000);
-        if (buf.u32 == 0) {
-            return true;
-        }
+        switch (ntrcard::state.status) {
+            case ntrcard::Status::RAW: {
+                if (sendAndVerifyRawInit()) {
+                    old_cart = false;
+                    return true;
+                } else if (trySecureInit(BlowfishKey::NTR)
+                    || trySecureInit(BlowfishKey::B9RETAIL)
+                    || trySecureInit(BlowfishKey::B9DEV)) {
+                    old_cart = true;
+                    return true;
+                }
 
-        // ok, we go the hard way
-        if (trySecureInit(BlowfishKey::NTR)
-            || trySecureInit(BlowfishKey::B9RETAIL)
-            || trySecureInit(BlowfishKey::B9DEV)) {
-            old_cart = true;
-            return true;
+                logMessage(LOG_ERR, "r4isdhc: init from RAW failed");
+                return false;
+            }
+            case ntrcard::Status::KEY2: {
+                return sendAndVerifyKey2Init();
```

Comment:
this case cannot detect new cart type. and old cart injecting will be wrong.

Reply:
> this case cannot detect new cart type.

New carts support both 0x66 in KEY2 and 0x68 in RAW.

> and old cart injecting will be wrong.

Why?